### PR TITLE
dbeaver/pro#2045 Improve gcp fs folders support

### PR DESCRIPTION
**Pull request by E1izabeth**
Please check creation and deletion for empty and non-empty folders.

### CODE REVIEW

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/fs/nio/EFSNIOFolder.java [hunk 1]
```diff
@@ -49,7 +55,8 @@ public void create(boolean force, boolean local, IProgressMonitor monitor) throw
 
     public void create(int updateFlags, boolean local, IProgressMonitor monitor) throws CoreException {
         try {
-            Files.createDirectory(getNioPath());
+            Path result = Files.createFile(getNioPath().resolve(EMPTY_FOLDER_FAKE_FILE_NAME));
```
Comment (serge-rider) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/fs/nio/EFSNIOFolder.java]: You cannot do it for all file systems. We need it only for GCP

**Review by serge-rider — CHANGES_REQUESTED** [review]

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/fs/nio/EFSNIOFolder.java (hunk 1 above)
Comment (E1izabeth) [reply plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/fs/nio/EFSNIOFolder.java]: Then I need this class to not be final

Comment (serge-rider) [reply plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/fs/nio/EFSNIOFolder.java]: I think it is better to use some flag in FS provider (DBFVirtualFileSystem? or maybe additional interface)

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/fs/DBFVirtualFileSystem.java [hunk 2]
```diff
@@ -62,6 +62,10 @@ public interface DBFVirtualFileSystem extends Closeable {
 
     default void refreshRoots(DBRProgressMonitor monitor) throws DBException {}
 
+    default boolean isFoldersExists() {
```
Comment (serge-rider) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/fs/DBFVirtualFileSystem.java]: Let's rename it to supportsEmptyFolders

**Review by serge-rider — CHANGES_REQUESTED** [review]

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/fs/nio/EFSNIOFolder.java [hunk 3]
```diff
@@ -66,7 +79,14 @@ public void createLink(URI location, int updateFlags, IProgressMonitor monitor)
 
     public void delete(boolean force, boolean keepHistory, IProgressMonitor monitor) throws CoreException {
         try {
-            Files.delete(getNioPath());
+            if (getRoot().getRoot().getFileSystem().supportsEmptyFolders()) {
+                Files.delete(getNioPath());
+            } else {
+                // We can't delete pseudo-folder, so this is a workaround
```
Comment (serge-rider) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/fs/nio/EFSNIOFolder.java:85]: Looks dangerous and unsafe.
Can we delete only our FOLDER_EMULATION_FILE_NAME?

**Review by serge-rider — CHANGES_REQUESTED** [review]

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/fs/nio/EFSNIOFolder.java (hunk 3 above)
Comment (E1izabeth) [reply plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/model/fs/nio/EFSNIOFolder.java:85]: Then folder will not be deleted on delete action

**Comment by MashaKorax** [comment]
verified

**Review by serge-rider — APPROVED** [review]

**Pull request merged**